### PR TITLE
Add handling of REBAR_CACHE_DIR env variable

### PR DESCRIPTION
**Pull request by x0id**
The REBAR_CACHE_DIR should be separated between releases to avoid unexpected dialyzer errors caused by cached files from previous "rebar3 dialyzer" runs.
By default rebar3 uses ~/.cache/rebar3, this caused problems in different kerl environments.

# Description

Handling of the REBAR_CACHE_DIR variable was added the same way as it was done for REBAR_PLT_DIR.

Closes #&lt;issue&gt;.

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/kerl/kerl/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by paulo-ferraz-oliveira** [comment]
@jadeallenx, I'd like to have #452 approved and merged before I go for #454, then #455, then this one, as I'm sure there'll be conflicts.

@x0id, thanks for this. A couple of comments:

* would you mind porting this to `fish` and `csh`?
  * if you do mind, would you create a task for this port to be done at a later date? (then maybe I can tackle it)?
* maybe you'll want to wait on the previous action I mentioned - approve + merge
* I think this solution is acceptable, though I wonder what'll happen to e.g. plugins that use `rebar3`'s cache directly (I didn't look into `rebar3` to see how the path to the folder is exposed - but I know for sure there are plugins that depend on it, and it'd be a pity to break local/CI compatibility)
  * in that case, if `rebar3`'s public functions don't take this variable into account (which I would say is oversight, if it happens) maybe we could open an issue there, too?
* this change is lacking tests - one could e.g. call the activate script on two different installations and make sure the `$absdir/.cache/rebar3` folders are populated, have different content

**Comment by x0id** [comment]
> @jadeallenx, I'd like to have #452 approved and merged before I go for #454, then #455, then this one, as I'm sure there'll be conflicts.
> 
> @x0id, thanks for this. A couple of comments:
> 
> * would you mind porting this to `fish` and `csh`?

what do you mean by that? I added lines to all variants of activate/deactivate including `fish` and `csh` I believe

>   
>   * if you do mind, would you create a task for this port to be done at a later date? (then maybe I can tackle it)?
> * maybe you'll want to wait on the previous action I mentioned - approve + merge
> * I think this solution is acceptable, though I wonder what'll happen to e.g. plugins that use `rebar3`'s cache directly (I didn't look into `rebar3` to see how the path to the folder is exposed - but I know for sure there are plugins that depend on it, and it'd be a pity to break local/CI compatibility)

What I could suggest - making that variable optional. I'll think about it.

>   
>   * in that case, if `rebar3`'s public functions don't take this variable into account (which I would say is oversight, if it happens) maybe we could open an issue there, too?

Not sure I'm aware of these details, if you have an example let me know, but I agree with the point.

> * this change is lacking tests - one could e.g. call the activate script on two different installations and make sure the `$absdir/.cache/rebar3` folders are populated, have different content

Is there any part of `kerl` code that is for testing? Sorry, I didn't look too deep so far, so I can't say I understand what is the testing strategy in this project.

**Comment by paulo-ferraz-oliveira** [comment]
> what do you mean by that? I added lines to all variants of activate/deactivate including fish and csh I believe

Right you are. The GitHub diff UI is so small I only saw the `bash` changes... good stuff!


> What I could suggest - making that variable optional. I'll think about it.

I'm Ok as-is, though, as long as `rebar3` honours it where it exposes the global cache folder... (later) I checked and it seems it does.

```erlang
    State5 = case os:getenv("REBAR_CACHE_DIR") of
                false ->
                    State4;
                CachePath ->
                    rebar_state:set(State4, global_rebar_dir, CachePath)
            end,
```

then later...

```erlang
global_cache_dir(Opts) ->
    Home = home_dir(),
    rebar_opts:get(Opts, global_rebar_dir, filename:join([Home, ".cache", "rebar3"])).
```

which I'd say is the "official" exposed function.

> Is there any part of kerl code that is for testing? Sorry, I didn't look too deep so far, so I can't say I understand what is the testing strategy in this project.

Look in the GitHub actions file, under `.github`. I can help you set something up if you're stuck...

**Comment by paulo-ferraz-oliveira** [comment]
So what's missing now is:

* rebase on `master` once #454, and #455 get merged
* add tests to GitHub CI

**Comment by jadeallenx** [comment]
Once we get this in I'll do a new release <3

**Comment by x0id** [comment]
I have rebased the PR onto the latest master after adding a couple of cleanup-improving updates.
Will work on the tests...

**Comment by x0id** [comment]
I added `tests/activate_test.sh` script for testing sh-based activate and kerl_deactovate cleanup. This is a basic test. It may be improved/extended later.
I have also tried to work on the fish version and suddenly realized that the fish-activate/deactivate is far from perfect. The cleanup code just didn't work out of the box - maybe nobody tested it.
My suggestion is to run CI, verify what has been done so far, and release it as is if no problems are found. I can work on validating/improving/testing fish (and csh) versions later in a separate branch/PR.

**Comment by jadeallenx** [comment]
> I have also tried to work on the fish version and suddenly realized that the fish-activate/deactivate is far from perfect. The cleanup code just didn't work out of the box - maybe nobody tested it.

Undoubtedly true - unless it’s zsh or bash it doesn’t get a lot of love. 

> My suggestion is to run CI, verify what has been done so far, and merge/release it as is if no problems are found. 

ok that seems reasonable 

> I can work on validating/improving/testing fish (and csh) versions later in a separate branch/PR.

that would be very welcome- thanks!

File: tests/activate_test.sh
Comment (paulo-ferraz-oliveira) [inline tests/activate_test.sh:49]:
```suggestion
[ -d "$existing_dir" ] || exit 1
```

File: tests/activate_test.sh
Comment (paulo-ferraz-oliveira) [inline tests/activate_test.sh:33]: It would probably be better to name the variables at the start of the function instead of using $1, $2, ... I already have an [issue to do that](https://github.com/kerl/kerl/issues/453) in other scopes, but we should prevent new occurrences 😄

File: tests/activate_test.sh
Comment (paulo-ferraz-oliveira) [inline tests/activate_test.sh:43]: This (if I'm reading it properly) proves the env is different, which is A step, but don't we also want to look for specific variables?

**Comment by paulo-ferraz-oliveira** [comment]
Looking good. ❤️ 2 minor changes requested, and a code update suggestion added.

> I can work on validating/improving/testing fish (and csh) versions later in a separate branch/PR

Agreed. Would you open an issue for this, referencing your comment, in that case? Thanks. Maybe we can even try do see what an AI would do for "convert this bash to that fish" and then adapt? For tests too, but that's "for the future".

Tests are failing and so is the Lint action. I can help with the lint, if you want, by doing it locally and proposing changes. For the tests, there's very little information, so I'm actually not sure what's breaking: maybe in case of error we could also output the diff?

**Comment by x0id** [comment]
I have fixed lint errors, activate-test integration in CI, and corrected activate/deactivate code slightly.
Now all tests went green (in my account).

File: tests/activate_test.sh
Comment (x0id) [reply tests/activate_test.sh:38]: I'll keep that in mind for the next phase. I have a plan to improve/extend the test.

**Comment by paulo-ferraz-oliveira** [comment]
Looks good to me. I'll wait for @jadeallenx.

@x0id, is the stuff you plan to do identified in the Issues? Or not interesting to have there (e.g. for pre-discussion before a pull request)?

**Review by jadeallenx — APPROVED** [review]

**Comment by jadeallenx** [comment]
Let's ship it :) Thanks!

**Comment by x0id** [comment]
> @x0id, is the stuff you plan to do identified in the Issues? Or not interesting to have there (e.g. for pre-discussion before a pull request)?

Tried to briefly describe in https://github.com/kerl/kerl/issues/458.
I have a [branch for the fish version](https://github.com/x0id/kerl/tree/fix-fish-scripts) already, still WIP.

**Comment by paulo-ferraz-oliveira** [comment]
Perfect, stuff, thanks much.

**Pull request merged**
